docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+S have this HardAffinityExpression in their PodSpec, then the HardAffinityExpression will block the first
+pod of the service from ever scheduling, since it is only allowed to run in a zone with another pod from
+the same service. And of course that means none of the pods of the service will be able
+to schedule. This problem *only* applies to hard affinity, not soft affinity or hard or soft
+anti-affinity. There are at least two ways to solve this problem
+* short-term: when a controller creates pods from a PodTemplate, for exactly *one* of those
+pods, it should omit any hard affinity rules that select the pods of that PodTemplate.
+* long-term/speculative: controllers could present the scheduler with a group of pods from
+the same PodTemplate as a single unit. This would allow two things: (1) the scheduler could ignore
+problematic hard affinity rules rather than requiring the controller to handle it,
+and (2) the scheduler could do proper [gang scheduling](https://github.com/kubernetes/kubernetes/issues/16845)
+since it could receive an entire gang simultaneously.
+
+### Anti-affinity
+
+* **Spread the pods of this service S across nodes and zones**:
+`{{PodSelector: <selector that matches S's pods>, TopologyKey: "node"}, {PodSelector: <selector that matches S's pods>, TopologyKey: "zone"}}`
+
+* **Don't co-locate pods of this service with pods from service "evilService"**:
+`{PodSelector: selector that matches evilService's pods, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods**:
+`{PodSelector: empty, TopologyKey: "node"}`
+
+* **Don't co-locate pods of this service with any other pods except other pods of this service**:
+Assuming pods from the service have some label `{key=service, value=S}`:
+`{PodSelector: "service" NotIn "S", TopologyKey: "node"}`
+Note that this works because `"service" NotIn "S"` matches pods with no key "service"
+as well as pods with key "service" and a corresponding value that is not "S."
+
+## Algorithm
+
+An example algorithm a scheduler might use to implement affinity and anti-affinity rules is as follows.
+There are certainly more efficient ways to do it; this is just intended to demonstrate that the API's
+semantics are implementable.
+
+Terminology definition: We say a pod P is "feasible" on a node N if P meets all of the scheduler
+predicates for scheduling P onto N. Note that there are various reasons why a pod that is correctly
+scheduled onto some node might become infeasible on that node later; for example,
+some labels might have changed on the node (thus violating some constraint of an already-running pod)
+or some labels might have changed on a pod (thus violating some AffinitySelector of an already-running pod).
+We assume pods that become infeasible after they are bound to a node may be allowed to continue running; the
+various policies and mechanisms for handling this scenario are outside the scope of this document, and
```

Comment:
Yes, that would be a reasonable theoretical semantic for hard affinities.  But I suspect that users will commonly blow their leg off with that, no mind shoot themselves in the foot.  At least if the terminations are not done very carefully.

The main concern I have is accidental cascading terminations. e.g. A->B->C->D.  If the user terminates D (or it just crashes), then the system automatically also terminates C then B then A.  That in itself is not totally bad, except for the fact that there's no easy way to tell that they're all going to be deleted by looking at D.  It's actually quite difficult to trace back through the dependency graph in the general case to figure out what the effect would be of terminating a given Pod.  To make matters worse, it's quite easy to concoct a not entirely unrealistic scenario where the termination of a single Pod (through either explicit user termination or even just a crash) causes all the other Pods in a cluster to be automatically terminated by the control plane (e.g. A (same host) B (same cluster) C).  This is a stable configuration until the (potentially single) C pod terminates, after which the system would automatically terminate (potentially thousands of) B and then A pods. 

The more I think about it, the more it seems sensible to make it the responsibility of a replication controller (or something like it) to (optionally) terminate pods whose hard constraints are no longer met.  RC's know about all of the replicas in a replica set, and can terminate and relaunch pods with violated hard constraints in a controlled fashion (much like a rolling update).  The recreated pods would be scheduled onto nodes such that the constraints are once again satisfied.

Does that sound better?